tests/nsrt_learning/strips_learning/test_gen_to_spec_learner.py

```diff
@@ -846,3 +868,63 @@ def test_keep_effect_adding_new_variables():
             assert seg is segmented_traj[1]
             assert sub == {potato_x0: potato3}
+
+
+@pytest.mark.parametrize("val", [0.0, 1.0])
+def test_multi_pass_backchaining(val):
+    """Test that the BackchainingSTRIPSLearner does multiple passes of
+    backchaining, which is needed to ensure harmlessness."""
+    utils.reset_config({"segmenter": "atom_changes"})
+    # Set up the types, objects, and, predicates.
+    dummy_type = Type("dummy_type",
+                      ["feat1", "feat2", "feat3", "feat4", "feat5"])
+    dummy = dummy_type("dummy")
+    A = Predicate("A", [], lambda s, o: s[dummy][0] > 0.5)
+    B = Predicate("B", [], lambda s, o: s[dummy][1] > 0.5)
+    C = Predicate("C", [], lambda s, o: s[dummy][2] > 0.5)
+    D = Predicate("D", [], lambda s, o: s[dummy][3] > 0.5)
+    E = Predicate("E", [], lambda s, o: s[dummy][4] > 0.5)
+    predicates = {A, B, C, D, E}
+
+    # Create the necessary options and actions.
+    Pick = utils.SingletonParameterizedOption("Pick",
+                                              lambda s, m, o, p: None,
+                                              types=[]).ground([], [])
+    pick_act = Action([], Pick)
+    Place = utils.SingletonParameterizedOption("Place",
+                                               lambda s, m, o, p: None,
+                                               types=[]).ground([], [])
+    place_act = Action([], Place)
+
+    # Create trajectories.
+    s10 = State({dummy: [1.0, 0.0, 0.0, 0.0, 0.0]})
+    s11 = State({dummy: [1.0, 1.0, 1.0, 0.0, 0.0]})
+    s12 = State({dummy: [1.0, 0.0, 1.0, 1.0, 1.0]})
+    traj1 = LowLevelTrajectory([s10, s11, s12], [pick_act, place_act], True, 0)
+    goal1 = {GroundAtom(D, [])}
+    task1 = Task(s10, goal1)
+
+    s20 = State({dummy: [1.0, 1.0, 0.0, 0.0, val]})
+    s21 = State({dummy: [1.0, 0.0, 0.0, 1.0, 1.0]})
+    traj2 = LowLevelTrajectory([s20, s21], [place_act], True, 1)
+    goal2 = {GroundAtom(D, []), GroundAtom(E, [])}
+    task2 = Task(s20, goal2)
+
+    s30 = State({dummy: [1.0, 1.0, val, 0.0, 0.0]})
+    s31 = State({dummy: [1.0, 0.0, 1.0, 1.0, 1.0]})
+    traj3 = LowLevelTrajectory([s30, s31], [place_act], True, 2)
+    goal3 = {GroundAtom(C, []), GroundAtom(D, [])}
+    task3 = Task(s30, goal3)
+
+    ground_atom_trajs = utils.create_ground_atom_dataset([traj1, traj2, traj3],
+                                                         predicates)
+    segmented_trajs = [segment_trajectory(traj) for traj in ground_atom_trajs]
+
+    # Now, run the learner on the three demos.
+    learner = _MockBackchainingSTRIPSLearner([traj1, traj2, traj3],
+                                             [task1, task2, task3],
+                                             predicates,
+                                             segmented_trajs,
+                                             verify_harmlessness=True)
+    # Running this automatically checks that harmlessness passes.
+    _ = learner.learn()
```

Comment:
let's we verify the outputted PNADs are what we would expect? i was too lazy to do that when i wrote the test, but i think we should add the check in now...